### PR TITLE
[prototype] sensors can request backfills via a new class that is constructed via RunRequest

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
I tried a different approach to adding backfill support to sensors.
Rather than putting the attributes to run a backfill in the `RunRequest` like in https://github.com/dagster-io/dagster/pull/22681, this PR adds a new class `BackfillDaemonRequest` that contains just the attributes that are allowed for backfills. The intention is that this class will remain private and that the public API will look like this 
```python
RunRequest.for_asset_slices(...) 
``` 

I think this approach has a couple benefits:
- clear separation of supported parameters. In #22681 we have to do a lot of checking of parameter combinations to ensure that asset check keys, run keys, partition key, etc are not used when requesting a backfill. 
- Supports backfill specific attributes. You can set a title and a description for a backfill. Maybe we should add this to runs too, but the separation of parameters lets us support these attributes for backfills without adding even more invalid parameter combinations
- The sensor definition and sensor daemon code are more legible. We can check the type of the run request to see how it should be handled. In #22681 we had to either to expensive computations over the partitions in an asset graph subset, or check which of `asset_selection` and `asset_graph_subset` was set to know how to handle the request. We may still want to do some work to determine which kind of request we make when the user calls `RunRequest.from_asset_slices` but that work will happen once at construction time, rather than several times in the sensor execution. it'll also prevent having potentially expensive logic in a function that seems like it should be fast `requires_backfill_daemon`

One concern is how this will affect type checking. If a user writes this 
```python 
@sensor(...)
def my_sensor(context) -> RunRequest:
   ...
   return RunRequest.from_asset_slices(...)
```
there will be a linting error since `BackfillDaemonRequest` is what's actually returned.

### CODE REVIEW

**Comment by jamiedemaria** [comment]
* **#22827** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22827?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jamiedemaria and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/run_request.py [hunk 1]
```diff
@@ -111,8 +114,68 @@ def __new__(
         )
 
 
+class IRunRequest(ABC):
+    @property
+    def requires_backfill_daemon(self) -> bool:
+        return False
+
+
+(
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/run_request.py:123]: remove

File: python_modules/dagster/dagster/_core/definitions/run_request.py [hunk 2]
```diff
@@ -111,8 +114,68 @@ def __new__(
...
+        ("asset_check_keys", PublicAttr[Optional[Sequence[AssetCheckKey]]]),
+    ],
+)
+
+
+@whitelist_for_serdes
+class MultiRunRequest(
+    IRunRequest,
+    NamedTuple(
+        "_RunRequest",
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/run_request.py:139]: _MultiRunRequest

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 3]
```diff
@@ -807,7 +809,10 @@ def evaluate_tick(self, context: "SensorEvaluationContext") -> "SensorExecutionD
         if not result or result == [None]:
             skip_message = "Sensor function returned an empty result"
         elif len(result) == 1:
-            item = check.inst(result[0], (SkipReason, RunRequest, DagsterRunReaction, SensorResult))
+            item = check.inst(
+                result[0],
+                (SkipReason, RunRequest, MultiRunRequest, DagsterRunReaction, SensorResult),
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:814]: IRunRequest

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 4]
```diff
@@ -870,14 +875,42 @@ def evaluate_tick(self, context: "SensorEvaluationContext") -> "SensorExecutionD
...
-            for run_request in self.resolve_run_requests(
-                run_requests, context, self._asset_selection, dynamic_partitions_requests
-            )
-        ]
+
+        run_requests_to_resolve = []
+        resolved_run_requests = []
+        for run_request in run_requests:
+            if run_request.requires_backfill_daemon:
+                asset_selection = check.not_none(
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:883]: can make these vars a better name. `asset_selection` -> `sensor_asset_selection`. `asset_keys` ->  `run_request_asset_keys` maybe

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 5]
```diff
@@ -870,14 +875,42 @@ def evaluate_tick(self, context: "SensorEvaluationContext") -> "SensorExecutionD
...
-            )
-            for run_request in self.resolve_run_requests(
-                run_requests, context, self._asset_selection, dynamic_partitions_requests
-            )
-        ]
+
+        run_requests_to_resolve = []
+        resolved_run_requests = []
+        for run_request in run_requests:
+            if run_request.requires_backfill_daemon:
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:882]: check pyright for this. might need to do a `isinstance(run_request, MultiRunRequest)` instead

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 6]
```diff
@@ -1084,13 +1115,13 @@ def _wrapped_fn(context: SensorEvaluationContext):
         raw_evaluation_result = fn(**context_param, **resource_args_populated)
 
         def check_returned_scalar(scalar):
-            if isinstance(scalar, (SkipReason, RunRequest, SensorResult)):
+            if isinstance(scalar, (SkipReason, RunRequest, SensorResult, MultiRunRequest)):
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:1118]: IRunRequest

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 7]
```diff
@@ -1084,13 +1115,13 @@ def _wrapped_fn(context: SensorEvaluationContext):
...
                 return scalar
             elif scalar is not None:
                 raise Exception(
                     f"Error in sensor {sensor_name}: Sensor unexpectedly returned output "
-                    f"{scalar} of type {type(scalar)}.  Should only return SkipReason or "
-                    "RunRequest objects."
+                    f"{scalar} of type {type(scalar)}.  Should only return SkipReason, "
+                    "RunRequest, or MultiRunRequest objects."
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:1124]: remove MultiRunRequst - user would only interact with RunRequest

File: python_modules/dagster/dagster/_daemon/sensor.py [hunk 8]
```diff
@@ -663,16 +666,26 @@ def _evaluate_sensor(
...
-            context=context,
-            instance=instance,
-            external_sensor=external_sensor,
-            workspace_process_context=workspace_process_context,
-            submit_threadpool_executor=submit_threadpool_executor,
-            sensor_debug_crash_flags=sensor_debug_crash_flags,
-        )
+        single_runs_to_handle = []
+        for run_request in sensor_runtime_data.run_requests:
+            if run_request.requires_backfill_daemon:
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_daemon/sensor.py:671]: check pyright. might need to do `isintance(run_requst, MultiRunRequst)`

File: python_modules/dagster/dagster/_daemon/sensor.py [hunk 9]
```diff
@@ -747,6 +760,26 @@ def _handle_dynamic_partitions_requests(
...
+    multi_run_request: MultiRunRequest,
+    instance: DagsterInstance,
+    context: SensorLaunchContext,
+) -> None:
+    backfill_id = make_new_backfill_id()
+    instance.add_backfill(
+        PartitionBackfill.from_asset_graph_subset(
+            backfill_id=backfill_id,
+            dynamic_partitions_store=instance,
+            backfill_timestamp=pendulum.now("UTC").timestamp(),
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_daemon/sensor.py:773]: remove pendulum

File: python_modules/dagster/dagster/_core/definitions/run_request.py [hunk 10]
```diff
@@ -276,6 +321,24 @@ def partition_key_range(self) -> Optional[PartitionKeyRange]:
...
+        title: Optional[str],
+        description: Optional[str],
+    ) -> BackfillDaemonRequest:
+        """Constructs a BackfillDaemonRequest from an AssetGraphSubset. When processed by the sensor
+        daemon, this will launch a backfill instead of a run.
+
+        Note: This constructor is intentionally left private since AssetGraphSubset is not part of the
+        public API. Other constructor methods will be public
+        """
+        return BackfillDaemonRequest(
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/run_request.py:338]: In this constructor method (and other future constructors) we could do some introspection on the provided asset_graph_subset to determine if the request can be executed in a single run (and if so return a RunRequest)

I'm inclined to punt on doing this for now since it's not simple to do this introspection efficiently. i think it could also make sense to have it always emit backfills (even if it could be a single run) during the initial rollout since DA will be the first user. If it's always a backfill, then the backfills page becomes the source of truth of what was launched while the new runs page is developed, rather than having some backfills and some single runs

figuring out how to determine if a request should be a single run or a backfill efficiently will be important for future backfill and run consolidation work, so i do plan to spend some time looking in to how to do this, i just dont think it should block this pr

File: python_modules/dagster/dagster/_core/definitions/run_request.py [hunk 11]
```diff
@@ -111,8 +114,50 @@ def __new__(
         )
 
 
+class IRunRequest(ABC):
+    pass
+
+
+@whitelist_for_serdes
+class BackfillDaemonRequest(
+    IRunRequest,
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/run_request.py:123]: I haven't encountered using a parent class and NamedTuple before. not sure if it is a python sin

**Comment by jamiedemaria** [comment]
@OwenKephart @gibsondan I tried a different approach to supporting backfills from sensors. I think i like the way this turned out more than #22681 except for the type hint stuff i mentioned in the description. lmk what you think! i'm going to spend some time thinking through both options and then will aim to settle on one by EOD friday/early monday

**Review by OwenKephart — COMMENTED** [review]
I hear you on the desire to organize things differently from the first PR, but I'm pretty wary of this IRunRequest approach -- typically an Interface sort of class would define some shared properties between things that implement it. The "more legit" pattern here would be to come up with some shared base class that's named "ExecutionRequest" or something, but the concern there is just that that sort of thing is pretty committal.

My next thought was that we could compose this differently, and just name `BackfillDaemonRequest` to `<Something>Args`, and toss that in as an optional field to `RunRequest`, but my concern there is that this also feels committal. e.g. we very well might decide in the future to allow a title / description for "regular" RunRequests. It's also quite hard to name the `<Something>` bit, as anything with the name `BackfillDaemonX` feels prescriptive (what if we decide to fulfill these without the BackfillDaemon in the future?).

Anyway, my two cents is that the other approach feels a bit safer / more future-proof. There is some gross stuff involved with validating the combinations of parameters, but I think that's less scary than adding a new serialized object like this. In terms of the perf stuff, calling `subset.size` is typically not an issue, it's just `len(subset.get_asset_partitions())` that's scary.

I also think it's fine if the initial version of the `requires_backfill_daemon` is as simple as checking if the AssetGraphSubset is None, but we'll want to adjust that fairly soon after (but again I'm not too worried about perf stuff there)

**Comment by jamiedemaria** [comment]
yeah i hear you on the interface being empty, it's not great. my hope was that some stuff would shake out in the implementation that would make sense to put there, but alas. Can you explain a bit more why the interface feels more committal than adding the parameter?

**Comment by OwenKephart** [comment]
re: the interface being more committal, the main things are:

1. It requires updating the types of a fair number of call sites, some of which are user-facing (that issue you note with RunRequest.X() not resulting in something of type `RunRequest` is fairly rough).
2. The main one in my mind is that this requires a whole new type of serialized object, which we need to create a name for. The sole purpose of this serialized object is to distinguish the execution pattern of the given request (daemon vs. traditional), which is aligned with our current-day constraints, but is not aligned with our future goals or the user-facing mental-model. i.e. this distinction should eventually become invisible to the user and potentially be completely unified at the infra level. So at that point in time, having two separate objects here would no longer make sense, but we'd be stuck supporting both of these objects. Finally, even in a world where we _don't_ fully unify the infrastructure (i.e. we still need two separate code paths for processing runs of certain types), I think we would still want to provide a way for users to create "traditional" runs targeting an `AssetGraphSubset` (as this paves the way for runs that target assets with different PartitionsDefinitions).

So TL;DR is that I think adding the parameter to RunRequest is likely to happen regardless of what infrastructure unification we end up doing, but having two separate execution paths is not necessarily going to be around forever and even if it is we might want to have the freedom to name it something different / reimagine the scenarios where we need to invoke its behavior vs. not.

**Comment by jamiedemaria** [comment]
discussed in standup - typing challenges of this PR deemed too nasty, going to go ahead with https://github.com/dagster-io/dagster/pull/22681 but hide the `asset_graph_subset` in  `RunRequest` in a `kwarg` to avoid the API ergonomic issues with that approach
